src/modules/dwm.cpp

```
@@ -0,0 +1,493 @@
+
+        // Initialize layout symbol
+        update_layout_label(m_bar_mon->layout.symbol.cur);
+
+        // This event is only needed to update the layout label
+        m_ipc->on_layout_change = [this](const dwmipc::LayoutChangeEvent& ev) { on_layout_change(ev); };
+        m_ipc->subscribe(dwmipc::Event::LAYOUT_CHANGE);
+        m_log.info("%s: Initialized layout label", name());
+      }
+
+      // These events are only necessary to update the focused window title
+      if (m_title_label) {
+        update_title_label();
+        m_ipc->on_client_focus_change = [this](const dwmipc::ClientFocusChangeEvent& ev) {
+          this->on_client_focus_change(ev);
+        };
+        m_ipc->subscribe(dwmipc::Event::CLIENT_FOCUS_CHANGE);
+
+        m_ipc->on_focused_title_change = [this](const dwmipc::FocusedTitleChangeEvent& ev) {
+          this->on_focused_title_change(ev);
+        };
+        m_ipc->subscribe(dwmipc::Event::FOCUSED_TITLE_CHANGE);
+        m_log.info("%s: Initialized title label", name());
+      }
+
+      if (m_floating_label) {
+        update_floating_label();
+        m_ipc->on_focused_state_change = [this](const dwmipc::FocusedStateChangeEvent& ev) {
+          this->on_focused_state_change(ev);
+        };
+        m_ipc->subscribe(dwmipc::Event::FOCUSED_STATE_CHANGE);
+        m_log.info("%s: Initialized floating label", name());
+      }
+
+      // This event is for keeping track of the currently focused monitor
+      m_ipc->on_monitor_focus_change = [this](const dwmipc::MonitorFocusChangeEvent& ev) {
+        this->on_monitor_focus_change(ev);
+      };
+      m_ipc->subscribe(dwmipc::Event::MONITOR_FOCUS_CHANGE);
+        m_log.info("%s: Subscribed to monitor focus change", name());
+    } catch (const dwmipc::IPCError& err) {
+      throw module_error(err.what());
+    }
+  }
+
+  void dwm_module::stop() {
+    try {
+      m_log.info("%s: Disconnecting from socket", name());
+      m_ipc.reset(nullptr);
+    } catch (...) {
```

We shouldn't just swallow all exceptions. We should at least print an error or warning message.

---

I agree, another lazy copy of the i3 module.

---

The i3 module doesn't always follow best practices ;)

---

See b24ce89d5d055273ee0637df166db2d775cc2b44